# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/2d/PathSkia.cpp
# Commit: 20b46afb225f
# Full Hash: 20b46afb225f7b0076e3fc0c000ce609ac79e0ba
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2023-04-11 09:27:51
# Regressor Bug: 1821512
# File Overlap Count: 1
# Description:
#   Bug 1821512 - Fix Skia usage for API changes. r=jrmuizel
#   
#   This just tries to address fairly random changes in the Skia API and correct
#   our usage of it in Moz2D and some other places.
#   
# ==============================================================================

diff -r b531f604ea4c -r 20b46afb225f gfx/2d/PathSkia.cpp
--- a/gfx/2d/PathSkia.cpp	Tue Apr 11 05:00:04 2023 +0000
+++ b/gfx/2d/PathSkia.cpp	Tue Apr 11 05:00:05 2023 +0000
@@ -7,7 +7,7 @@
 #include "PathSkia.h"
 #include "HelpersSkia.h"
 #include "PathHelpers.h"
-#include "skia/src/core/SkDraw.h"
+#include "skia/include/core/SkPathUtils.h"
 #include "skia/src/core/SkGeometry.h"
 
 namespace mozilla::gfx {
@@ -30,9 +30,9 @@
 void PathBuilderSkia::SetFillRule(FillRule aFillRule) {
   mFillRule = aFillRule;
   if (mFillRule == FillRule::FILL_WINDING) {
-    mPath.setFillType(SkPath::kWinding_FillType);
+    mPath.setFillType(SkPathFillType::kWinding);
   } else {
-    mPath.setFillType(SkPath::kEvenOdd_FillType);
+    mPath.setFillType(SkPathFillType::kEvenOdd);
   }
 }
 
@@ -129,10 +129,24 @@
   return SkPathContainsPoint(mPath, aPoint, aTransform);
 }
 
-float ComputeResScaleForStroking(const Matrix& aTransform) {
+bool PathSkia::GetFillPath(const StrokeOptions& aStrokeOptions,
+                           const Matrix& aTransform, SkPath& aFillPath,
+                           const Maybe<Rect>& aClipRect) const {
+  SkPaint paint;
+  if (!StrokeOptionsToPaint(paint, aStrokeOptions)) {
+    return false;
+  }
+
   SkMatrix skiaMatrix;
   GfxMatrixToSkiaMatrix(aTransform, skiaMatrix);
-  return SkDraw::ComputeResScaleForStroking(skiaMatrix);
+
+  Maybe<SkRect> cullRect;
+  if (aClipRect.isSome()) {
+    cullRect = Some(RectToSkRect(aClipRect.ref()));
+  }
+
+  return skpathutils::FillPathWithPaint(mPath, paint, &aFillPath,
+                                        cullRect.ptrOr(nullptr), skiaMatrix);
 }
 
 bool PathSkia::StrokeContainsPoint(const StrokeOptions& aStrokeOptions,
@@ -142,15 +156,11 @@
     return false;
   }
 
-  SkPaint paint;
-  if (!StrokeOptionsToPaint(paint, aStrokeOptions)) {
+  SkPath strokePath;
+  if (!GetFillPath(aStrokeOptions, aTransform, strokePath)) {
     return false;
   }
 
-  SkPath strokePath;
-  paint.getFillPath(mPath, &strokePath, nullptr,
-                    ComputeResScaleForStroking(aTransform));
-
   return SkPathContainsPoint(strokePath, aPoint, aTransform);
 }
 
@@ -169,15 +179,12 @@
     return Rect();
   }
 
-  SkPaint paint;
-  if (!StrokeOptionsToPaint(paint, aStrokeOptions)) {
+  SkPath fillPath;
+  if (!GetFillPath(aStrokeOptions, aTransform, fillPath)) {
     return Rect();
   }
 
-  SkPath result;
-  paint.getFillPath(mPath, &result);
-
-  Rect bounds = SkRectToRect(result.computeTightBounds());
+  Rect bounds = SkRectToRect(fillPath.computeTightBounds());
   return aTransform.TransformBounds(bounds);
 }
 